File: 2021_12_24_advent_of_code_in_fsharp/AdventOfCode/AdventOfCode2021/Day04.fs
module AdventOfCode2021.Day4

open Common
open System
open System.Linq

type Cell = { Value: int; mutable IsChecked: bool }

type Board = { Cells: Cell [,]; mutable IsChecked: bool }

type Data = { Numbers: int list; Boards: Board list }

[<Literal>]
let BoardSize = 5

let loadData (input: string) =
    let lines = input.Split Environment.NewLine
    let boards =
        lines
        |> Seq.skip 2
        |> Seq.filter (fun line -> line.Length > 0)
        |> Seq.chunkBySize BoardSize
        |> Seq.map (
            Seq.map (parseNumbers ' ')
            >> Seq.map (Array.map (fun n -> { Value = n; IsChecked = false }))
            >> array2D
            >> (fun cells -> { Cells = cells; IsChecked = false })
        )
        |> Seq.toList
    { Numbers = (lines.[0] |> parseNumbers ',' |> Array.toList); Boards = boards }


let getRowsAndColumns (b: Board) =
    [ 0 .. BoardSize - 1 ] |> Seq.collect (fun index -> [ b.Cells.[index, *]; b.Cells.[*, index] ])

let checkCell n (cell: Cell) =
    cell.IsChecked <- cell.Value = n
    cell.IsChecked

let checkBoard n (board: Board) =
    board.IsChecked <-
        board
        |> getRowsAndColumns
        |> Seq.map (fun line -> line |> Seq.filter (fun c -> c.IsChecked || checkCell n c) |> Seq.length)
        |> Seq.exists ((=) BoardSize)
    board.IsChecked

let findWinningBoards (data: Data) =
    data.Numbers
    |> Seq.collect (fun n ->
        data.Boards |> Seq.filter (fun b -> not b.IsChecked) |> Seq.filter (checkBoard n) |> Seq.map (fun b -> n, b))

let countScoreForBoard (board: Board) =
    board.Cells
    |> Enumerable.OfType<Cell>
    |> Seq.choose (fun cell ->
        match cell with
        | { IsChecked = true } -> None
        | { Value = value } -> Some value)
    |> Seq.sum

let puzzle1 (input: string) =
    let data = loadData input
    let n, b = data |> findWinningBoards |> Seq.head
    let result = n * countScoreForBoard b
    result |> string


let puzzle2 (input: string) =
    let data = loadData input
    let n, b = findWinningBoards data |> Seq.last
    let result = n * countScoreForBoard b
    result |> string

File: 2022_04_23_advent_of_code_in_csharp/AdventOfCode/AdventOfCode2021/Day04.cs
using static AdventOfCode.Common;
using static AdventOfCode.Option;
using static System.Linq.Enumerable;

namespace AdventOfCode.AdventOfCode2021;

static class Day4
{
    private const int BoardSize = 5;

    public record Cell(int Value, bool IsChecked)
    {
        public bool IsChecked { get; set; } = IsChecked;
    }

    public record Board(Cell[,] Cells, bool IsChecked)
    {
        public bool IsChecked { get; set; } = IsChecked;
    }

    public record Data(int[] Numbers, Board[] Boards);

    private static T[,] Array2D<T>(IEnumerable<IEnumerable<T>> rows)
    {
        var array = rows.Select(row => row.ToArray()).ToArray();
        var rowCount = array.Length;
        var columnCount = array[0].Length;
        var result = new T[rowCount, columnCount];
        for (var r = 0; r < rowCount; ++r)
        {
            var row = array[r];
            for (var c = 0; c < columnCount; ++c)
            {
                result[r, c] = row[c];
            }
        }
        return result;
    }

    public static Data LoadData(string input)
    {
        var lines = input.Split(Environment.NewLine);
        var boards =
            lines
            .Skip(2)
            .Where(line => line.Length > 0)
            .Chunk(BoardSize)
            .Select(chunk => new Board(chunk.Select(line => ParseNumbers(' ', line).Select(n => new Cell(n, false))).Pipe(Array2D), false))
            .ToArray();
        return new Data(ParseNumbers(',', lines[0]).ToArray(), boards);
    }

    static IEnumerable<Cell[]> GetRowsAndColumns(Board board) =>
        Range(0, BoardSize).SelectMany(x =>
            new[]
            {
                Range(0, BoardSize).Select(y => board.Cells[x,y]).ToArray(),
                Range(0, BoardSize).Select(y => board.Cells[y,x]).ToArray(),
            }
        );

    static bool CheckCell(int n, Cell cell) => cell.IsChecked = cell.Value == n;

    static bool checkBoard(int n, Board board) =>
        board.IsChecked = GetRowsAndColumns(board)
            .Select(line => line.Count(c => c.IsChecked || CheckCell(n, c)))
            .Any(count => count == BoardSize);


    static IEnumerable<(int, Board)> FindWinningBoards(Data data) =>
        data.Numbers.SelectMany(
            n => data.Boards.Where(b => !b.IsChecked && checkBoard(n, b)).Select(b => (n, b)));

    static int CountScoreForBoard(Board board) =>
        board.Cells
            .OfType<Cell>()
            .Choose(cell => cell switch
            {
                { IsChecked: true } => None<int>(),
                { Value: var value } => Some(value),
            })
            .Sum();


    public static string Puzzle1(string input)
    {
        var data = LoadData(input);
        var (n, b) = data.Pipe(FindWinningBoards).First();
        var result = n * CountScoreForBoard(b);
        return result.ToString();
    }

    public static string Puzzle2(string input)
    {
        var data = LoadData(input);
        var (n, b) = data.Pipe(FindWinningBoards).Last();
        var result = n * CountScoreForBoard(b);
        return result.ToString();
    }
}